### PR TITLE
UI Manager and UI / UX for Powerup Icons

### PR DESCRIPTION
**Pull request by seanlowjk**
# Summary 

1) Added `UIManager` to handle all UI/UX related stuff in the scene or the game 
2) Created `UIIconAnimator` and `UIIconTimer` to ensure that the Animations properly fade out. 
3) Added extra tidbits to ensure that the Icons appear when in contact with the powerups and added the magic sound effect. All of these can be found in `Assets\Scenes\Gyms\GYM_UI\GYM_HUD_3.unity`

## UIManager

As mentioned earlier, the reason why we have a UIManager is to manage all things related to UI UX! For now, there will be just 1 method to activate the relavant power-up icon. 

## UIIconAnimator and UIIconTimer 

To ensure the timer and the fade out works, these components are added. 

## Extra tidbits 

You will need to add the SFX to the `AudioManager` in order for it to be recognized. Else, you will not be able to play the sound. 😨 

# To-Dos: 

- [x] Gather Feedback about the animations for the powerup icons! All can be found in `Assets\Scenes\Gyms\GYM_UI\GYM_HUD_3.unity`
- [x] Rename `PowerUpPanelHandler` to `PowerUpUIHandler`
- [x] Shift Logic to include Enumerables 
- [x] Move Animation Scripts to Actual Mecanim Animatior

### CODE REVIEW

File: Assets/Scripts/UI/Powerups/PowerUpPanelHandler.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        if (m_JumpPowerUpIconPrefab != null)
+        {
+            GameObject jumpIcon = Instantiate(m_JumpPowerUpIconPrefab) as GameObject;
+            jumpIcon.transform.SetParent(this.gameObject.transform, false);
+        }
+    }
+
+    // Simple call to Activate Jump Icon on Power Up Panel. 
+    public void ActivateSpeedIcon()
```
Comment (heze8) [inline Assets/Scripts/UI/Powerups/PowerUpPanelHandler.cs]: Can this methods be done in one method?

File: Assets/Scripts/UI/UIManager.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class UIManager : Singleton<UIManager>
+{
+    [SerializeField]
+    private GameObject m_powerUpHandler;
+
+    public static string JumpSignal = "Jump";
```
Comment (heze8) [inline Assets/Scripts/UI/UIManager.cs]: Are these constants? can use an enum here if they are

Comment (seanlowjk) [reply Assets/Scripts/UI/UIManager.cs]: Truth

File: Assets/Scripts/UI/Powerups/PowerUpPanelHandler.cs (hunk 1 above)
Comment (seanlowjk) [reply Assets/Scripts/UI/Powerups/PowerUpPanelHandler.cs]: Truth hor, okay will fix in next commit 💯

File: Assets/Scripts/UI/Powerups/Icons/UIIconAnimator.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    private void DecreaseAlphaValue(Outline outline, float decrementFactor)
+    {
+        if (outline != null)
+        {
+            var outlineColor = outline.effectColor;
+            outlineColor.a -= decrementFactor * Time.deltaTime;
+            outline.effectColor = outlineColor;
+        }
+    }
```
Comment (Eclmist) [inline Assets/Scripts/UI/Powerups/Icons/UIIconAnimator.cs]: I feel like these can all be driven by an animator for better artistic control. How about creating a animator for the powerup icons and passing a single 0-1 float value into the animator and let it control all the opacity, size, fill amount, etc?

Comment (seanlowjk) [reply Assets/Scripts/UI/Powerups/Icons/UIIconAnimator.cs]: Hmmm, that actually makes sense. Then in the future, when we reuiqre different kinds of poweups and skills which are time-bases, we can use the animator for greater use!

File: Assets/Scripts/UI/Powerups/PowerUpPanelHandler.cs (hunk 1 above)
Comment (seanlowjk) [reply Assets/Scripts/UI/Powerups/PowerUpPanelHandler.cs]: Fixed!

File: Assets/Scripts/UI/UIManager.cs (hunk 2 above)
Comment (seanlowjk) [reply Assets/Scripts/UI/UIManager.cs]: Fixed!

File: Assets/Scripts/UI/Powerups/Icons/UIIconAnimator.cs (hunk 3 above)
Comment (seanlowjk) [reply Assets/Scripts/UI/Powerups/Icons/UIIconAnimator.cs]: Fixed!

**Comment by seanlowjk** [comment]
image

Added the Decrementer Animation with a variable to make it easier to set duration of icon :P

File: Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            {
+                activatedIcon = Instantiate(m_JumpPowerUpIconPrefab) as GameObject;
+            }
+            break;
+        case UIPowerUpSignals.SPEED_SIGNAL:
+            if (m_SpeedPowerUpIconPrefab != null)
+            {
+                activatedIcon = Instantiate(m_SpeedPowerUpIconPrefab) as GameObject;
+            }
+            break;
```
Comment (tcgj) [inline Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs]: Single line if statements by convention we're doing
```C#
if (condition)
    doSomething;```

File: Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        default: 
+            return;
+        }
+
+        if (activatedIcon != null)
+        {
+            activatedIcon.transform.SetParent(this.gameObject.transform, false);
+        }
```
Comment (tcgj) [inline Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs]: Same thing here. Also, if you want to access the transform, it's just `transform`. No need to do `this.gameobject.transform`

File: Assets/Scripts/PowerUps/PowerUpBase.cs [hunk 6]
```diff
@@ -19,6 +19,17 @@ void Update()
...
+            PlayPickUpSound();
+            HandlePowerUp(player);
+            Destroy(gameObject);
+        }
+    }
+
+    private void PlayPickUpSound()
+    {
+        AudioManager.m_Instance.PlaySound("MAGIC_Powerup", 1.2f, 1.2f);
         m_TimeOfActivation = Time.time;
```
Comment (tcgj) [inline Assets/Scripts/PowerUps/PowerUpBase.cs]: I think this was a merge issue. The lines your pr is trying to add here should have been removed when we added caleb's item pickups.

File: Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs (hunk 4 above)
Comment (seanlowjk) [reply Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs]: Epic!

File: Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs (hunk 5 above)
Comment (seanlowjk) [reply Assets/Scripts/UI/Powerups/PowerUpUIHandler.cs]: okay fast gameeeeeeeeeeeee

File: Assets/Scripts/PowerUps/PowerUpBase.cs (hunk 6 above)
Comment (seanlowjk) [reply Assets/Scripts/PowerUps/PowerUpBase.cs]: hmmm... interesting...

ok shall fix that nao

**Pull request merged**